pkg/deploy/controller/deploymentconfig/factory.go

```diff
@@ -8,76 +8,221 @@ import (
 	"k8s.io/kubernetes/pkg/client/cache"
 	"k8s.io/kubernetes/pkg/client/record"
 	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	kcontroller "k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/framework"
 	"k8s.io/kubernetes/pkg/runtime"
-	"k8s.io/kubernetes/pkg/util/flowcontrol"
 	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
-	"k8s.io/kubernetes/pkg/watch"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
 
 	osclient "github.com/openshift/origin/pkg/client"
-	controller "github.com/openshift/origin/pkg/controller"
 	deployapi "github.com/openshift/origin/pkg/deploy/api"
 )
 
-// DeploymentConfigControllerFactory can create a DeploymentConfigController which obtains
-// DeploymentConfigs from a queue populated from a watch of all DeploymentConfigs.
-type DeploymentConfigControllerFactory struct {
-	// Client is an OpenShift client.
-	Client osclient.Interface
-	// KubeClient is a Kubernetes client.
-	KubeClient kclient.Interface
-	// Codec is used to encode/decode.
-	Codec runtime.Codec
+const (
+	// FullDeploymentConfigResyncPeriod means we'll attempt to reconcile deploymentconfigs
+	// every two minutes.
+	FullDeploymentConfigResyncPeriod = 2 * time.Minute
+	// We must avoid creating new replication controllers until the {replication controller,pods} store
+	// has synced. If it hasn't synced, to avoid a hot loop, we'll wait this long between checks.
+	StoreSyncedPollPeriod = 100 * time.Millisecond
+)
+
+// NewDeploymentConfigController creates a new DeploymentConfigController.
+func NewDeploymentConfigController(dcInformer, rcInformer, podInformer framework.SharedIndexInformer, oc osclient.Interface, kc kclient.Interface, codec runtime.Codec) *DeploymentConfigController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartRecordingToSink(kc.Events(""))
+	recorder := eventBroadcaster.NewRecorder(kapi.EventSource{Component: "deploymentconfig-controller"})
+
+	c := &DeploymentConfigController{
+		dn: oc,
+		rn: kc,
+
+		queue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		recorder: recorder,
+		codec:    codec,
+	}
+
+	c.dcStore.Indexer = dcInformer.GetIndexer()
+	dcInformer.AddEventHandler(framework.ResourceEventHandlerFuncs{
+		AddFunc:    c.addDeploymentConfig,
+		UpdateFunc: c.updateDeploymentConfig,
+		DeleteFunc: c.deleteDeploymentConfig,
+	})
+
+	c.rcStore.Indexer = rcInformer.GetIndexer()
+	rcInformer.AddEventHandler(framework.ResourceEventHandlerFuncs{
+		AddFunc:    c.addReplicationController,
+		UpdateFunc: c.updateReplicationController,
+		DeleteFunc: c.deleteReplicationController,
+	})
+
+	c.rcStoreSynced = rcInformer.HasSynced
+	c.podStoreSynced = podInformer.HasSynced
```

Comment:
removed; it is used in the upstream controller because they are updating pod labels at some point.